compiler/src/iree/compiler/InputConversion/Common/QuantizedConvToConv.cpp

```diff
@@ -0,0 +1,320 @@
+
+  Value init = makeLinalgInit(rewriter, loc, conv);
+  return rewriter
+      .create<linalg::GenericOp>(
+          loc, init.getType(), ValueRange{conv, sum}, ValueRange{init},
+          affineMaps, iterators,
+          [=](OpBuilder &b, Location loc, ValueRange args) {
+            Value mul = b.create<arith::MulIOp>(loc, args[1], zp);
+            Value sum = b.create<arith::SubIOp>(loc, args[0], mul);
+            b.create<linalg::YieldOp>(loc, sum);
+          })
+      .getResult(0);
+}
+
+// Reduce the input value along the reduction dimensions
+Value reduceDims(PatternRewriter &rewriter, Location loc, Value val,
+                 Type accETy, ArrayRef<bool> is_reduction) {
+  auto context = val.getContext();
+  RankedTensorType ty = val.getType().cast<RankedTensorType>();
+
+  llvm::SmallVector<int64_t> staticSizes;
+  SmallVector<Value> dynSizes;
+  for (int i = 0, s = is_reduction.size(); i < s; i++) {
+    if (is_reduction[i]) continue;
+
+    staticSizes.push_back(ty.getDimSize(i));
+    if (ty.isDynamicDim(i)) {
+      dynSizes.push_back(rewriter.create<tensor::DimOp>(loc, val, i));
+    }
+  }
+
+  // Create a zero-filled accumulator.
+  Value initAcc =
+      rewriter.create<tensor::EmptyOp>(loc, staticSizes, accETy, dynSizes);
+  Value zeroInt =
+      rewriter.create<arith::ConstantIntOp>(loc, 0, accETy).getResult();
+  Value zeroAcc =
+      rewriter.create<linalg::FillOp>(loc, zeroInt, initAcc).getResult(0);
+
+  SmallVector<AffineExpr> filterExprs(ty.getRank());
+  SmallVector<AffineExpr> outputExprs;
+  SmallVector<StringRef> iterators;
+
+  for (int i = 0, s = ty.getRank(); i < s; i++) {
+    if (!is_reduction[i]) {
+      auto expr = rewriter.getAffineDimExpr(iterators.size());
+      iterators.push_back(getParallelIteratorTypeName());
+
+      outputExprs.push_back(expr);
+      filterExprs[i] = expr;
+    }
+  }
+
+  for (int i = 0, s = filterExprs.size(); i < s; i++) {
+    if (!filterExprs[i]) {
+      auto expr = mlir::getAffineDimExpr(iterators.size(), context);
+      iterators.push_back(getReductionIteratorTypeName());
+      filterExprs[i] = expr;
+    }
+  }
+
+  SmallVector<AffineMap> affineMaps{AffineMap::get(4, 0, filterExprs, context),
+                                    AffineMap::get(4, 0, outputExprs, context)};
+
+  return rewriter
+      .create<linalg::GenericOp>(
+          loc, zeroAcc.getType(), ValueRange{val}, ValueRange{zeroAcc},
+          affineMaps, iterators,
+          [=](OpBuilder &b, Location loc, ValueRange args) {
+            Value ext = b.create<arith::ExtSIOp>(loc, accETy, args[0]);
+            Value sum = b.create<arith::AddIOp>(loc, ext, args[1]);
+            b.create<linalg::YieldOp>(loc, sum);
+          })
+      .getResult(0);
+}
+
+// Pattern lowering conv_2d_nhwc_hwcf_q to conv_2d_nhwc_hwcf.
+//
+// This is implementing the math explained in Section 2.3 of
+// https://arxiv.org/abs/1712.05877.
+struct QuantizedConvToConv
```

Comment:
Do you see opportunities to share helpers or anything with https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/InputConversion/Common/QuantizedMatmulToMatmul.cpp ? Anything worth mentioning in a comment? Though, these two files being side by side in the same directory, makes the commonality discoverable.

Reply:
I had to write a more general reduction operation for the convolution case so I moved it to a utils file and reused across both passes.